File: rickroll.pl
#!/usr/bin/perl -w
use 5.010;
use strict;
use warnings;
use autodie;
use Method::Signatures;

use Dancer;
use Data::Dumper;

set port => 3001;
set logger => 'console';
set log => 'core';
set warnings => 1;

# The location of this script.
my $callback_url = "http://exobrain.pjf.id.au:3001/rickroll";

# Played on timeout.
my $goodbye = "Thank you for calling the National Rick Astley Hotline. Goodbye.";

# Tune selections

my $original = {
    description => "The Original",
    by => "Rick Astley",
    url => "https://dl.dropboxusercontent.com/u/9702672/music/01-NeverGonnaGiveYouUp.mp3",
};

# TODO: I originally made this a hash so "original" could be 9, but
# it's kinda a pain in the ass, so at some point we should make this
# an array with auto-numbering.
my %tunes = (
    1 => {
        # https://www.youtube.com/watch?v=b1WWpKEPdT4
        description => "Eight bit",
        by => "Kita Khyber",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/8-Bit%20Rick%20Roll.mp3",
    },
    2 => {
        # https://www.youtube.com/watch?v=q-9KqwCFDJs
        description => "Dubstep",
        by => "Crystalize",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick-Astley-Dubstep.mp3",
    },
    3 => {
        # Not sure who this is by
        # https://vimeo.com/64322245
        description => "Daft Punk",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick%20Roll%20Never%20Gonna%20Give%20You%20Up%20%28Daft%20Punk%20remix%29.mp3",
    },
    4 => {
        # Avicii, but Twilio gets confused by that.
        # https://www.youtube.com/watch?v=oT3mCybbhf0
        description => "Uh-vee-chee",
        by => "Nils",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/AVICII%20and%20RICK%20ASTLEY%20-%20Never%20Gonna%20Wake%20Up%20%28Mashup-Remix%29.mp3",
    },
    5 => {
        # https://www.youtube.com/watch?v=Eupg7rZ9AUY
        description => "Drum and bass",
        by => "Wave-shapers",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick%20Astley%20-%20Never%20Gonna%20Give%20You%20Up%20%28WAV35HAPERS%20Remix%29.mp3",
    },
    6 => {
        # https://www.youtube.com/watch?v=KykFbfCMizo
        description => "E.D.M.",
        by => "Riot", # R!OT
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick%20Astley%20-%20Never%20Gonna%20Give%20You%20Up%20%28R%21OT%20Remix%29.mp3",
    },
    7 => {
        # https://www.youtube.com/watch?v=GL-8XuoxuaQ
        description => "Metal",
        by => "Andy Rehfeldt",
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick%20Astley-Never%20Gonna%20Give%20You%20Up%28Metal%20Version%29.mp3",
    },
    8 => {
        # https://www.youtube.com/watch?v=snC4ZtW9dHI
        description => "Nirvana",
        by => "i. v. lad e. o", # ivladeo
        url => "https://dl.dropboxusercontent.com/u/9702672/music/Rick%20Astley%20%20%20Nirvana%20Mashup%20%20%20Never%20gonna%20give%20your%20teen%20spirit%20up.mp3",
    },

    9 => $original,

    # Zero is menu! ^_^
);

# Prep work, prepare our menu.

my $menu = "Welcome to the national Rick Astley hotline. You may make your selection at any time.\n";

foreach my $key (sort keys %tunes) {
    my $artist = $tunes{$key}{by} || "unknown";
    $menu .= "To listen to $tunes{$key}{description} by $artist, press $key.\n";
}

$menu .= "To hear these options again, press zero.\n";
$menu .= "If you do not wish to be rick-rolled, please hang up now.";

# This is a regexp, as there may be a leading path in our URL.
any qr{.*/rickroll} => sub {

    my $selection = param "Digits";

    # No selection is a first-time caller. Give them the classic rick.
    if (not defined $selection) {
        return PlayTune($original);
    }

    # Play a tune selection, if we have one.
    elsif (my $tune = $tunes{$selection}) {
        return PlayTune($tune);
    }

    # Otherwise play our menu.
    return Menu();

};

any '/sms' => sub {
    warn Dumper params();
    return Twilio("");
};

dance;

func PlayTune($tune) {
    return Twilio( Gather($callback_url, Play($tune->{url}), Say($menu) ), Say($goodbye) );
}

sub Menu {
    return Twilio( Gather($callback_url, Say($menu) ), Say($goodbye) );
}

sub Say    { return qq{<Say>@_</Say>}                     }
sub Play   { return qq{<Play loop="1">@_</Play>}          }
sub Record { return qq{<Record action="@_" />}            }

func Pause($seconds = 1) {
    return qq{<Pause length="$seconds"/>}
}

func Gather($url!, @args) { 
    return qq{<Gather numDigits="1" timeout="10" action="$url">@args</Gather>} 
}

func Redirect($url!, $method = "GET") {
    return qq{<Redirect method="$method">/$url</Redirect>\n};
}

sub Twilio {
    return qq{<?xml version="1.0" encoding="UTF-8" ?>\n<Response>@_</Response>};
}

File: rickroll.py
from flask import Flask, request
app = Flask(__name__)

from twilio import twiml
from six.moves.urllib.parse import quote
import yaml

with open('songs.yaml') as f:
    yaml_content = yaml.safe_load(f.read())

tunes = [yaml_content[song] for song in yaml_content['remixes']]
tunes.insert(0, None)  # Zero is always the menu.
tunes.append(yaml_content['original'])  # Make sure we add in the original.

# Played on timeout
goodbye = "Thank you for calling the National Rick Astley Hotline. Goodbye."

def play_tune(tune):
    """Takes a tune dictionary, returns a TwiML response that plays it."""

    # If somehow we're called with no argument, which could happen if we had
    # a short menu or multi-key input.
    if tune is None:
        return play_menu()

    response = twiml.Response()

    # By calling functions on gather, digits can be pressed during the song
    # playback *and* the menu afterwards.
    gather = response.gather(numDigits=1, timeout=10)
    gather.play(yaml_content['url_base'] + quote(tune['filename']))
    gather.say(generate_menu())
    
    # Our goodbye triggers after gather times out.
    response.say(goodbye)

    return response

def play_menu():
    """Plays the menu"""

    response = twiml.Response()

    gather = response.gather(numDigits=1, timeout=10)
    gather.say(generate_menu())

    response.say(goodbye)

    return response

@app.route("/", methods = ['GET','POST'])
def original():

    selection = request.values.get('Digits')

    # Without a selection, play the original
    if selection is None:
        return str(play_tune(yaml_content['original']))

    # With a selection try to turn into an index we can use.
    # There may be a '#' or '*', so we'll default to '0'
    # (our menu) if we fail to parse as an int
    try:
        selection = int(selection)
    except ValueError:
        selection = 0

    # Zero is always our menu.
    if selection == 0:
        return str(play_menu())

    # Otherwise load the song they want, with a default of the original
    # song if they select something outsie our array bounds.
    tune = yaml_content['original']

    try: tune = tunes[selection]
    except Exception: pass

    return str(play_tune(tune))

def generate_menu():
    menu = "Welcome to the national Rick Astley hotline. You may make your selection at any time.\n"

    for idx, song in enumerate(tunes):

        if song is None:
            continue

        menu += "To listen to {}".format(song['title'])

        if 'artist' in song:
            menu += " by {}".format(song['artist'])

        menu += ", press {}.\n".format(idx)

    menu += "To hear these options again, press zero.\n"
    menu += "If you do not wish to be rick-rolled again, please hang up now."

    return menu

if __name__ == "__main__":
    app.run()
